---
title: "scMRD_VAF_comparison"
author: "TR"
date: "2023-04-27"
output: html_document
---

```{r}

library(tidyverse)
library(pals)
library(readxl)
library(ggpubr)

```
Read in VAF data
```{r}

VAF_data <- read_xlsx("MRD_v2_VAF_compare.xlsx")

```

Plot scMRD v. Bulk VAF for all mutations (detected by either bulk, scMRD, or both)
```{r}

sc_bulk_plot <-  ggplot(VAF_data, aes(x=scDNA_VAF, y=Bulk_VAF, fill = Gene)) +
                 geom_point(size=4, shape=21,alpha = 0.9, show.legend = TRUE)+
                 geom_abline(slope = 1, intercept = 0, linetype = "dashed")+
                      theme_classic()+
                        theme(plot.title = element_text(hjust = 0.5, size = 20),
                              axis.text.y = element_text(size=12, face = "bold"), 
                              axis.text.x = element_text(size=12, face = "bold"), 
                              axis.title.y = element_text(size = 14), 
                              axis.title.x = element_text(size = 14))+
                         scale_y_continuous(trans=scales::pseudo_log_trans(base = 10), limits = c(0,99))+
                         scale_x_continuous(trans=scales::pseudo_log_trans(base = 10), limits = c(0,99))+
                  xlab("Variant Allele Frequency (scMRD)")+
                  ylab("Variant Allele Frequency (Bulk)")+
                  ggtitle("scMRD v. Bulk VAF")


sc_bulk_plot

   #ggsave(plot = sc_bulk_plot, width = 7, height = 5, dpi = 500, filename = "MRD2_v2_VAF_compare.tiff", device = "tiff")

```

Only compare VAF of mutations detected by both bulk NGS and scMRD in the chunk below.
```{r}

detected_in_both <- VAF_data %>% #select mutations that were detected by both methods to do statistical comparison
                    dplyr::filter(Bulk_VAF > 0) %>%
                    dplyr::filter(scDNA_VAF > 0) %>% reshape2::melt(.)

paired_colors<- c("Bulk_VAF" = "#EA3546", "scDNA_VAF" = "#00A7E1")
my_comparisons <- list(c("Bulk_VAF", "scDNA_VAF"))


scMRD_Bulk_VAF_paired <- ggpaired(detected_in_both, x = "variable", y = "value", fill  = scales::alpha(paired_colors,0.7),
                                  line.color = "gray50", line.size = 0.5)+
                         geom_hline(yintercept = 0, linetype = "dashed")+
                                scale_fill_manual(values = )+
                                theme_classic()+
                                theme(axis.text.y = element_text(size = 12),
                                      axis.text.x = element_text(size = 12),
                                      axis.title.y = element_text(size = 14))+
                                ylab("Variant Allele Frequency (VAF)")+ 
                                stat_compare_means(comparisons = my_comparisons, 
                                                   method = "wilcox.test", 
                                                   paired = TRUE, 
                                                   label.y = 100, 
                                                   size = 4)

scMRD_Bulk_VAF_paired

```





